docs/madr/decisions/010-timeout-policy.md

```diff
@@ -0,0 +1,174 @@
+# MeshUpstreamTimeout policy
+
+* Status: ?
+
+Technical Story: [#4739](https://github.com/kumahq/kuma/issues/4739)
+
+## Context and Problem Statement
+
+[New policy matching MADR](https://github.com/kumahq/kuma/blob/master/docs/madr/decisions/005-policy-matching.md) introduces 
+a new approach how Kuma applies configuration to proxies. Rolling out strategy implies creating a new set of policies that 
+satisfy a new policy matching. Current MADR aims to define a `MeshUpstreamTimeout`.
+
+## Considered Options
+
+* Create `MeshUpstreamTimeout` policy
+
+## Decision Outcome
+
+TBD.
+
+## Solution
+
+### How Timeout policy can be configured right now
+This is the default timeout configuration in Kuma right now:
+
+```yaml
+ conf:
+    connectTimeout: 5s # all protocols
+    tcp: 
+      idleTimeout: 1h 
+    http:
+      requestTimeout: 15s 
+      idleTimeout: 1h
+      streamIdleTimeout: 30m
+      maxStreamDuration: 0s
+```
+
+Why idle timeout can be configured on both TCP and HTTP connections?
```

Comment:
I also can't see any valid use case for this.

Reply:
They could be configured on both TCP and HTTP connections because they could have a different values. Practically they're also pretty different (from kuma doc):
```
TCP
// IdleTimeout is defined as the period in which there are no bytes sent
// or received on either the upstream or downstream connection
```
```
HTTP
// IdleTimeout is the time at which a downstream or upstream connection
// will be terminated if there are no active streams
```
So TCP idle timeout starts counting when no bytes are sent in the connection, but HTTP idle timeout starts counting when all streams of the connection are closed. In order for streams to be closed `stream_idle_timeout` should be over. TCP idle timeout is closer to HTTP stream idle timeout.

Reply:
Does it mean that if I set `idleTimeout` for HTTP for 2h and don't modify TCP `idleTimeout`. There can be a situation that stream hangs in the middle and connection will be killed by TCP `idleTimeout` after 1h and http `idleTimeout` doesn't even start counting? 

Won't it be misleading for users? Wouldn't single `idleTimeout` be simpler and easier to configure correctly without taking too much freedom from users? Or do we even need http `idleTimeout` Do you see some good use cases for it?

Reply:
In this case, that makes sense to split idle timeouts and preserve the TCP section timeout

Reply:
@lobkovilya what do you think about it?

Reply:
In that case, we're losing the ability to create a global policy with different tcp and http idle timeouts:
```yaml
type: MeshTimeout
name: mt-1
mesh: mesh-1
spec:
  targetRef:
    kind: Mesh
  to:
    - targetRef:
        kind: Mesh
      default:
        tcp:
          idleTimeout: 1h
        http:
          idleTimeout: 2h       
```
And maybe this is okay for simplicity.

Also, I'm thinking now, that technically we can support `MeshSubset` kind in `to` section, but limited only to the `kuma.io/protocol` tag. So we'll rewrite the example above:
```yaml
type: MeshTimeout
name: mt-1
mesh: mesh-1
spec:
  targetRef:
    kind: Mesh
  to:
    - targetRef:
        kind: MeshSubset
        tags:
          kuma.io/protocol: tcp
      default:
        idleTimeout: 1h
    - targetRef:
        kind: MeshSubset
        tags:
          kuma.io/protocol: http
      default:
        idleTimeout: 2h   
```

Now the question is do we want to have `MeshSubset` in `to` with only 1 available `kuma.io/protocol` tag

Reply:
I vote for one top-level idleTimeout. Yes, technically we lose the capability of Global idle timeout for different protocols, but don't think this is important. We can introduce separate idle timeouts if needed.

It's harder to go from specific timeouts for protocols to one timeout.

Reply:
I agree with Jakub